spec/Classes.tex

```diff
@@ -663,73 +600,453 @@ \subsection{User-Defined Constructors}
-% - eliminate compiler-generated constructor when there is a user-defined one
-% This leads us to the "unlike a function call" semantics here -
-% to avoid duplicate invocation of fields' default initializers
-% (which otherwise would happen once during the object's default initialization
-% and the second time upon invoking the compiler-generated constructor).
-% But overall I do not feel that this refactoring is a win, especially
-% because we now need to define the magic (and I am not sure presently
-% it is defined formally enough).
-
-\begin{chapelexample}{defaultConstructor.chpl}
+\subsubsection{Field Initialization Versus Assignment}
+\label{Field_Initialization_Versus_Assignment}
+\index{classes!initializers!user-defined!init-vs-assign}
+
+Within the body of an initializer, the first use of a field as the left-hand
+side of an assignment statement will be considered initialization. Subsequent
+uses of the assignment operator on the field will invoke regular assignment as
+defined by the language.
+
+\begin{chapelexample}{fieldInitAssignment.chpl}
+The following example documents the difference between field initialization
+and field assignment.
+\begin{chapel}
+class PointDoubleX {
+  var x, y : real;
+
+  proc init(x: real, y: real) {
+    this.x = x;              // initialization
+    writeln("x = ", this.x); // use of initialized field
+    this.x = this.x * 2;     // assignment, use of initialized field
+
+    this.y = y;              // initialization
+  }
+}
+
+var p = new PointDoubleX(1.0, 2.0);
+\end{chapel}
+\begin{chapelpost}
+writeln(p);
+\end{chapelpost}
+\begin{chapeloutput}
+x = 1.0
+{x = 2.0, y = 2.0}
+\end{chapeloutput}
+The first statement in the initializer initializes field \chpl{x} to the value
+of the formal \chpl{x}. The second statement simply reads the value of the
+initialized field. The third statement reads the value of the field, doubles
+it, and {\em assigns} the result to the field \chpl{x}.
+\end{chapelexample}
+
+If a field is used before it is initialized, an compile-time error will be
+issued.
+
+\begin{chapelexample}{fieldUsedBeforeInitialized.chpl}
+In the following code:
+\begin{chapel}
+class Point {
+  var x, y : real;
+
+  proc init(x: real, y: real) {
+    writeln(this.x); // Error: use of uninitialized field!
+    this.x = x;
+    this.y = y;
+    writeln(this.y);
+  }
+}
+var p = new Point(1.0, 2.0);
+\end{chapel}
+\begin{chapelcompopts}
+--no-warnings
+\end{chapelcompopts}
+
+The first statement in the initializer reads the value of uninitialized field
+\chpl{x}, so the compiler will issue an error.
+
+\begin{chapelprintoutput}
+fieldUsedBeforeInitialized.chpl:5: error: field "x" used before it is initialized
+\end{chapelprintoutput}
+
+\end{chapelexample}
+
+\subsubsection{Omitting Field Initializations}
+\label{Omitting_Field_Initializations}
+\index{classes!initializers!user-defined!omitting}
+
+Some classes have many fields, some of which might have default values. Instead
```

Comment:
Maybe reword "Some classes have many fields,  some of which might have default values". Important part is default values, not many fields. 

Maybe "To support XYZ, the language allows ABC" would be a better structure.